src/palace/manager/sqlalchemy/session.py

```diff
@@ -0,0 +1,182 @@
+    if hasattr(obj, "__html__"):
+        return str(obj.__html__())
+
+    # Pass everything else off to Pydantic JSON encoder.
+    return pydantic_encoder(obj)
+
+
+def json_serializer(*args, **kwargs) -> str:
+    return json.dumps(*args, default=json_encoder, **kwargs)
+
+
+class SessionManager:
+    # A function that calculates recursively equivalent identifiers
+    # is also defined in SQL.
+    RECURSIVE_EQUIVALENTS_FUNCTION = "recursive_equivalents.sql"
+
+    @classmethod
+    def engine(cls, url: str | None = None, poolclass: type[Pool] | None = None):
+        url = url or Configuration.database_url()
+        return create_engine(
+            url,
+            echo=DEBUG,
+            json_serializer=json_serializer,
+            pool_pre_ping=True,
+            poolclass=poolclass,
+        )
+
+    @classmethod
+    def setup_event_listener(
+        cls, session: Session | sessionmaker
+    ) -> Session | sessionmaker:
+        event.listen(session, "before_flush", Listener.before_flush_event_listener)
+        return session
+
+    @classmethod
+    def sessionmaker(cls, url=None, session=None):
+        if not (url or session):
+            url = Configuration.database_url()
+        if url:
+            bind_obj = cls.engine(url)
+        elif session:
+            bind_obj = session.get_bind()
+            if not os.environ.get("TESTING"):
+                # If a factory is being created from a session in test mode,
+                # use the same Connection for all of the tests so objects can
+                # be accessed. Otherwise, bind against an Engine object.
+                bind_obj = bind_obj.engine
+        session_factory = sessionmaker(bind=bind_obj)
+        cls.setup_event_listener(session_factory)
+        return session_factory
+
+    @classmethod
+    def initialize_schema(cls, engine):
+        """Initialize the database schema."""
+        # Use SQLAlchemy to create all the tables.
+        Base.metadata.create_all(engine)
+
+    @classmethod
+    def session(cls, url, initialize_data=True, initialize_schema=True):
+        engine = cls.engine(url)
+        connection = engine.connect()
+        return cls.session_from_connection(connection)
+
+    @classmethod
+    def session_from_connection(cls, connection: Connection) -> Session:
+        session = Session(connection)
+        cls.setup_event_listener(session)
+        return session
+
+    @classmethod
+    def initialize_data(cls, session: Session):
+        # Check if the recursive equivalents function exists already.
+        query = (
+            select([literal_column("proname")])
+            .select_from(table("pg_proc"))
+            .where(literal_column("proname") == "fn_recursive_equivalents")
+        )
+        result = session.execute(query).all()
+
+        # If it doesn't, create it.
+        if not result:
+            resource_file = (
+                resources_dir("sqlalchemy") / cls.RECURSIVE_EQUIVALENTS_FUNCTION
```

Comment:
Only code changes here are to load this from the package, rather then manipulating file paths here.